### PR TITLE
fix(router): Allow redirects after an absolute redirect

### PR DESCRIPTION
**Pull request by atscott**
The router currently restricts all further redirects after an absolute redirect. Because there's no documented reason for _why_ this restriction is in place, I'm now deeming this unnecessary. Developers should not be restricted in this manner. Instead, configs that may have caused infinite redirects in the past should be updated to not be infinite. It is confusing to ignore configs with redirects after an absolute redirect occurred because it creates different matching rules depending on the whether an absolute redirect has happened or not.

For additional context on why I believe removing this restriction is
necessary, https://github.com/angular/angular/issues/13373 asks for allowing `redirectTo` to be a function. It
would make sense to allow this function to return a `UrlTree` like other
guards in the Router. When guards in the `Router` return `UrlTree`, they
cancel the current navigation and start a new one to re-do the route
matching. Since we're already in the router matching part, we don't need
to cancel the navigation. However, the restriction on absolute redirects
here then creates a weird situation where developers wouldn't see any
other redirects if they returned a `UrlTree` as an absolute redirect
from `redirectTo`.

BREAKING CHANGE: Absolute redirects no longer prevent further redirects. Route configurations may need to be adjusted to prevent infinite redirects where additional redirects were previously ignored after an absolute redirect occurred.

### CODE REVIEW

File: packages/router/src/recognize.ts [hunk 1]
```diff
@@ -62,51 +63,47 @@ export class Recognizer {
...
+      return {state: routeState, tree};
+    }));
   }
 
 
-  private match(tree: UrlTree) {
+  private match(rootSegmentGroup: UrlSegmentGroup): Observable<TreeNode<ActivatedRouteSnapshot>[]> {
+    if (ngDevMode) {
+      currentRedirectCount++;
+      if (currentRedirectCount > 100) {
```
Comment (AndrewKushnir) [inline packages/router/src/recognize.ts:94]: nit: should we extract this into a separate const (which would also allow to document this)?
```suggestion
      if (currentRedirectCount > MAX_ALLOWED_REDIRECTS) {
```

File: packages/router/test/apply_redirects.spec.ts [hunk 2]
```diff
@@ -1216,8 +1216,9 @@ describe('redirects', () => {
     it('should work when using absolute redirects (wildcard)', () => {
       checkRedirect(
           [
-            {path: '**', redirectTo: '/b(aux:c)'}, {path: 'b', component: ComponentB},
-            {path: 'c', component: ComponentC, outlet: 'aux'}
+            {path: 'b', component: ComponentB},
```
Comment (AndrewKushnir) [inline packages/router/test/apply_redirects.spec.ts:1219]: Should we add a test to verify that we stop redirecting once we hit the max redirects limit?

**Review by AndrewKushnir — APPROVED** [review]

**Review by AndrewKushnir — APPROVED** [review]
Reviewed-for: public-api

File: packages/router/src/recognize.ts (hunk 1 above)
Comment (atscott) [reply packages/router/src/recognize.ts:94]: Yep! Meant to do this but forgot before sending it out for review. Thanks.

**Review by dylhunn — APPROVED** [review]
reviewed-for: public-api

**Comment by atscott** [comment]
Removing `risk: medium` label. I've updated the approach to throw a dev mode error on infinite redirects but revert to previous behavior in production by blocking further redirects after the max has been reached.

**Comment by atscott** [comment]
[green TGP](https://fusion2.corp.google.com/presubmit/564508859/OCL:564508859:BASE:566312293:1695051418880:9adc3ace;groups=PossiblyNewlyFailing/targets)

**Comment by atscott** [comment]
caretaker note: Please merge and sync on its own. Also, I can't get all the CI jobs to pass - it's a new failure every time and none look related to this change.

**Comment by dylhunn** [comment]
This PR was merged into the repository by commit ce1b915868e654cdb679e9381db9d3bd3d68d5c4.

**Pull request closed without merging**